defrag/defrag.go

```diff
@@ -0,0 +1,293 @@
+// fragments, it will return nil and store whatever internal
+// information it needs to eventually defrag the packet.
+//
+// If the IPv4 layer is the last fragment needed to reconstruct
+// the packet, a new IPv4 layer will be returned, and will be set to
+// the entire defragmented packet,
+//
+// It use a map of all the running flows
+//
+// Usage example:
+//
+// func HandlePacket(in *layers.IPv4) err {
+//     defragger := ip4defrag.NewIPv4Defrag()
+//     in, err := defragger.DefragIPv4(in)
+//     if err != nil {
+//         return err
+//     } else if in == nil {
+//         return nil  // packet fragment, we don't have whole packet yet.
+//     }
+//     // At this point, we know that 'in' is defragmented.
+//     //It may be the same 'in' passed to
+//	   // HandlePacket, or it may not, but we don't really care :)
+//	   ... do stuff to 'in' ...
+//}
+//
+func (d *IPv4Defragmenter) DefragIPv4(in *layers.IPv4) (*layers.IPv4, error) {
+	// check if we need to defrag
+	if st := d.dontDefrag(in); st == true {
+		return in, nil
+	}
+	// perfom security checks
+	st, err := d.securityChecks(in)
+	if err != nil || st == false {
+		return nil, err
+	}
+
+	// ok, got a fragment
+	debug.Printf("defrag: got in.Id=%d in.FragOffset=%d in.Flags=%d\n",
+		in.Id, in.FragOffset*8, in.Flags)
+
+	// do we already has seen a flow between src/dst with that Id
+	ipf := newIPv4Flow(in)
+	var fl *fragmentList
+	var exist bool
+	d.Lock()
+	fl, exist = d.ipFlows[ipf]
+	if !exist {
+		debug.Printf("defrag: creating a new flow\n")
+		fl = new(fragmentList)
+		d.ipFlows[ipf] = fl
+	}
+	d.Unlock()
+	// insert, and if final build it
+	out, err2 := fl.insert(in)
+
+	// at last, if we hit the maximum frag list len
+	// without any defrag success, we just drop everything and
+	// raise an error
+	if out == nil && fl.List.Len()+1 > IPv4MaximumFragmentListLen {
+		d.Lock()
+		fl = new(fragmentList)
+		d.ipFlows[ipf] = fl
+		d.Unlock()
+		return nil, fmt.Errorf("defrag: Fragment List hits its maximum"+
+			"size(%d), without sucess. Flushing the list",
+			IPv4MaximumFragmentListLen)
+	}
+
+	// if we got a packet, it's a new one, and he is defragmented
+	if out != nil {
+		return out, nil
+	}
+	return nil, err2
+}
+
+// DiscardOlderThan forgets all packets without any activity since
+// time t. It returns the number of FragmentList aka number of
+// fragment packets it has discarded.
+func (d *IPv4Defragmenter) DiscardOlderThan(t time.Time) int {
+	var nb int
+	d.Lock()
+	for k, v := range d.ipFlows {
+		if v.LastSeen.Before(t) {
+			nb = nb + 1
+			delete(d.ipFlows, k)
+		}
+	}
+	d.Unlock()
+	return nb
+}
+
+// dontDefrag returns true if the IPv4 packet do not need
+// any defragmentation
+func (d *IPv4Defragmenter) dontDefrag(ip *layers.IPv4) bool {
+	// don't defrag packet with DF flag
+	if ip.Flags&layers.IPv4DontFragment != 0 {
+		return true
+	}
+	// don't defrag not fragmented ones
+	if ip.Flags == 0 && ip.Id == 0 {
```

Comment:
Hum... After some research, I found RFC 6864 and yes, Id field is/would be used for other stuff.